### DISCUSSION TITLE
CORS on Heroku?

### DISCUSSION THREAD

**Discussion by LukeKosner** [fastapi/fastapi#3867] [Questions] [answered]
Hey all!

I'm trying to build a FastAPI backend for my Svelte app. I had previously encountered a CORS error on Cross-Origin requests (specifics below), but I could solve them with the native CORS for FastAPI. The Heroku-deployed version would not deploy when I used them, though. Are there any other CORS solutions I could use?

_specifics_
 I've deployed the backend with [@AkhileshNS/heroku-deploy](https://github.com/akhileshns/heroku-deploy).

The CORS errors I received in the console were like this:
Access to XMLHttpRequest at _____(url) from origin ____(url) has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.

Much Appreciated,
Luke

**Comment by Mause** [comment, marked as answer]
What do you mean by "The Heroku-deployed version would not deploy when I used them"?

**Reply by LukeKosner** [reply]
@Mause When I added the native FastAPI CORS the deployment to Heroku with GitHub Actions failed.

**Reply by LukeKosner** [reply]
I'm looking for other CORS solutions or methods.

**Reply by Mause** [reply]
> @Mause When I added the native FastAPI CORS the deployment to Heroku with GitHub Actions failed.

With what error?

**Reply by LukeKosner** [reply]
> > @Mause When I added the native FastAPI CORS the deployment to Heroku with GitHub Actions failed.
> 
> With what error?

@Mause 
Sorry for the back and forth. I'm new to Open Source. Here's the whole error log. This only happens with FastAPI CORS. Is there any other way of doing CORS?
Run akhileshns/heroku-deploy@v3.0.4
12
Created and wrote to ~./netrc
13
 ›   Warning: Our terms of service have changed: 
14
Successfully logged into heroku
15
 ›   https://dashboard.heroku.com/terms-of-service
16
Added git remote heroku
17
heroku: Press any key to open up the browser to login or q to exit: remote: Compressing source files... done.        
18
remote: Building source:        
19
remote: 
20
remote: -----> Building on the Heroku-20 stack        
21
remote: -----> Using buildpacks:        
22
remote:        1. https://github.com/lstoll/heroku-buildpack-monorepo        
23
remote:        2. heroku/python        
24
remote: -----> Monorepo app detected        
25
remote:       Copied web to root of app successfully        
26
remote: -----> App not compatible with buildpack: https://buildpack-registry.s3.amazonaws.com/buildpacks/heroku/python.tgz        
27
remote:        More info: https://devcenter.heroku.com/articles/buildpacks#detection-failure        
28
remote: 
29
remote:  !     Push failed        
30
remote: Verifying deploy...        
31
remote: 
32
remote: !	Push rejected to intersectbe.        
33
remote: 
34
To https://git.heroku.com/intersectbe.git
35
 ! [remote rejected] HEAD -> master (pre-receive hook declined)
36
error: failed to push some refs to 'https://git.heroku.com/intersectbe.git'
37

38
            Unable to push branch because the branch is behind the deployed branch. Using --force to deploy branch. 
39
            (If you want to avoid this, set dontuseforce to 1 in with: of .github/workflows/action.yml. 
40
            Specifically, the error was: Error: Command failed: git push heroku HEAD:refs/heads/master 
41
remote: Compressing source files... done.        
42
remote: Building source:        
43
remote: 
44
remote: -----> Building on the Heroku-20 stack        
45
remote: -----> Using buildpacks:        
46
remote:        1. https://github.com/lstoll/heroku-buildpack-monorepo        
47
remote:        2. heroku/python        
48
remote: -----> Monorepo app detected        
49
remote:       Copied web to root of app successfully        
50
remote: -----> App not compatible with buildpack: https://buildpack-registry.s3.amazonaws.com/buildpacks/heroku/python.tgz        
51
remote:        More info: https://devcenter.heroku.com/articles/buildpacks#detection-failure        
52
remote: 
53
remote:  !     Push failed        
54
remote: Verifying deploy...        
55
remote: 
56
remote: !	Push rejected to intersectbe.        
57
remote: 
58
To https://git.heroku.com/intersectbe.git
59
 ! [remote rejected] HEAD -> master (pre-receive hook declined)
60
error: failed to push some refs to 'https://git.heroku.com/intersectbe.git'
61

62
        
63
remote: Compressing source files... done.        
64
remote: Building source:        
65
remote: 
66
remote: -----> Building on the Heroku-20 stack        
67
remote: -----> Using buildpacks:        
68
remote:        1. https://github.com/lstoll/heroku-buildpack-monorepo        
69
remote:        2. heroku/python        
70
remote: -----> Monorepo app detected        
71
remote:       Copied web to root of app successfully        
72
remote: -----> App not compatible with buildpack: https://buildpack-registry.s3.amazonaws.com/buildpacks/heroku/python.tgz        
73
remote:        More info: https://devcenter.heroku.com/articles/buildpacks#detection-failure        
74
remote: 
75
remote:  !     Push failed        
76
remote:  !        
77
remote:  ! ## Warning - The same version of this code has already been built: fdd4c6872c459770bbee9197704a73546aba572c        
78
remote:  !        
79
remote:  ! We have detected that you have triggered a build from source code with version fdd4c6872c459770bbee9197704a73546aba572c        
80
remote:  ! at least twice. One common cause of this behavior is attempting to deploy code from a different branch.        
81
remote:  !        
82
remote:  ! If you are developing on a branch and deploying via git you must run:        
83
remote:  !        
84
remote:  !     git push heroku <branchname>:main        
85
remote:  !        
86
remote:  ! This article goes into details on the behavior:        
87
remote:  !   https://devcenter.heroku.com/articles/duplicate-build-version        
88
remote: 
89
remote: Verifying deploy...        
90
remote: 
91
remote: !	Push rejected to intersectbe.        
92
remote: 
93
To https://git.heroku.com/intersectbe.git
94
 ! [remote rejected] HEAD -> master (pre-receive hook declined)
95
error: failed to push some refs to 'https://git.heroku.com/intersectbe.git'
96
Error: Error: Command failed: git push heroku HEAD:refs/heads/master --force
97
remote: Compressing source files... done.        
98
remote: Building source:        
99
remote: 
100
remote: -----> Building on the Heroku-20 stack        
101
remote: -----> Using buildpacks:        
102
remote:        1. https://github.com/lstoll/heroku-buildpack-monorepo        
103
remote:        2. heroku/python        
104
remote: -----> Monorepo app detected        
105
remote:       Copied web to root of app successfully        
106
remote: -----> App not compatible with buildpack: https://buildpack-registry.s3.amazonaws.com/buildpacks/heroku/python.tgz        
107
remote:        More info: https://devcenter.heroku.com/articles/buildpacks#detection-failure        
108
remote: 
109
remote:  !     Push failed        
110
remote:  !        
111
remote:  ! ## Warning - The same version of this code has already been built: fdd4c6872c459770bbee9197704a73546aba572c        
112
remote:  !        
113
remote:  ! We have detected that you have triggered a build from source code with version fdd4c6872c459770bbee9197704a73546aba572c        
114
remote:  ! at least twice. One common cause of this behavior is attempting to deploy code from a different branch.        
115
remote:  !        
116
remote:  ! If you are developing on a branch and deploying via git you must run:        
117
remote:  !        
118
remote:  !     git push heroku <branchname>:main        
119
remote:  !        
120
remote:  ! This article goes into details on the behavior:        
121
remote:  !   https://devcenter.heroku.com/articles/duplicate-build-version        
122
remote: 
123
remote: Verifying deploy...        
124
remote: 
125
remote: !	Push rejected to intersectbe.        
126
remote: 
127
To https://git.heroku.com/intersectbe.git
128
 ! [remote rejected] HEAD -> master (pre-receive hook declined)
129
error: failed to push some refs to 'https://git.heroku.com/intersectbe.git

**Reply by LukeKosner** [reply]
As an idea, is there a way to manually add the headers that makes sense?

**Reply by frankie567** [reply]
From what I read from the error log you provided, it seems that it's more an error related to Heroku deployment rather than CORS. It doesn't seem able to deploy your application because the commit you are trying to deploy is behind the one currently on Heroku branch:

```
remote: ! Push rejected to intersectbe.
33
remote:
34
To https://git.heroku.com/intersectbe.git
35
! [remote rejected] HEAD -> master (pre-receive hook declined)
36
error: failed to push some refs to 'https://git.heroku.com/intersectbe.git'
37

38
Unable to push branch because the branch is behind the deployed branch. Using --force to deploy branch.
```

## Deploy your Heroku app manually

I understand you use a GitHub Action to help you with that; however, my advice would be first to try to do it yourself using the Heroku CLI and see if it makes a difference. Here is the official getting started tutorial: https://devcenter.heroku.com/articles/getting-started-with-python#set-up Basically:

If not already, you should define a proper `Procfile` at the root of your repository where you define the start command of your FastAPI app. Something like this:

```
web: uvicorn your_app_module:app --host 0.0.0.0 --port $PORT
```
Then:

1. Install the Heroku CLI on your computer
2. Authenticate the CLI with your Heroku account: `heroku login`
3. Attach your local Git repository with the remote Heroku branch: `heroku git:remote -a name-of-your-heroku-app`
4. Push your Git branch to Heroku : `git push heroku master` (or `git push heroku main` if your main branch is called `main`)$
5. The deploy will start and you'll see the output in your terminal
6. If you get the same error about push rejection, you can *force* it to override the current commit: `git push --force heroku master`

## About GitHub automatic deployment

As a side note, once you get this working, you don't need a GitHub Action to automatically deploy from GitHub to Heroku: they have a built-in support for this, you only have to connect your GitHub repository with your Heroku app: https://devcenter.heroku.com/articles/github-integration

**Reply by LukeKosner** [reply]
@frankie567 and @Mause Thank you so much for the help! I've solved my problem with the GitHub auto-deploys. I had to split my code up into polyrepos but otherwise it worked great. I'll watch this awesome repo to help others with similar issues!
